Introduction of cluster controller and agent

ckreibich opened the pull request:
This is the current state of the ongoing controller work. It provides a functional controller and agent, plus a prototypical `zeek-client` submodule'd in from https://github.com/zeek/zeek-client (currently in Python, for quicker hacking). Only a small subset of the API laid out in the design doc is supported, but it's enough to use the client to ship a cluster topology, have the controller push that to the agents, and launch the cluster.

In terms of review, I'd suggest focusing it on the following:
- Are the overall layout and naming okay? I put the agent and controller under `scripts/base/frameworks/cluster`, but it's not technically related to the current cluster framework. Still it felt like a good home there for now.
- The only functionality that intersects with what people normally do with Zeek is in the minor modifications to the Supervisor. It can now listen, supports passing env vars to the launched nodes, and uses a different boot file (`scripts/base/init-supervisor.zeek`) when running (with `-j`).  

If you want to tinker, I recommend the following:
- Configure this branch with `--enable-zeek-client` and install.
- Clone https://github.com/ckreibich/cc-exploration, which contains a small runnable setup. (This'll move to docker-compose soon, but so far you didn't really need one if you've built it locally anyway.) 
- In the clone, run `zeek -j ./test-a2c.zeek` (as root).  This launches Zeek with one agent and one controlller. "a2c" means the agent connects to the controller; you can also use `test-c2a.zeek` for the other direction.
- Explore `zeek-client`. Saying `zeek-client instances` should list the instances (machines, in the future) connected to the controller. To push a cluster layout, say `zeek-client set-config -l ./etc/cluster.cfg`.

rsmmr commented on scripts/base/frameworks/cluster/agent/boot.zeek:
```diff
@@ -0,0 +1,35 @@
+
+	local epi = ClusterAgent::endpoint_info();
+	local sn = Supervisor::NodeConfig($name=epi$id,
+		$scripts=vector("base/frameworks/cluster/agent/main.zeek"));
```
Suggest to name `main.zeek` differently as the normal convention is for `__load__.zeek` to pull in `main.zeek`, which isn't the case here. Maybe even use another directory level to separate things out into what's always loaded vs what not.

ckreibich replied:
Hah. Until two days ago it was "runtime.zeek". :-) Do you prefer that? I figured there should be a `main.zeek` somewhere, since it's what people are trained to look for the essence of the functionality, and that's closest.

rsmmr reviewed (commented):
I've so far focused on the parts on integration into Zeek, not the new script functionality. I did try the `cc-exploration` scripts and that worked all great.

The main overall thing I'm wondering about is the structure of how the controller functionality is loaded & activated. Right now, it looks like it this:

- `zeek -j` bootstraps with `init-supervisor.zeek`, which then loads `cluster/agent` and `cluster/controller`.
- `zeek -j` is otherwise essetially bare mode because it doesn't pull in `init-default`.
- `cluster/{agent, controller}`  are no-ops unless `ClusterController::enable` gets set.
- Loading `test-a2c.zeek` sets `ClusterController::enable` and thereby enables the new functionality.
- When enabled, `cluster/{agent, controller}` spawn supervised processes loading `{agent,controller}/main.zeek` respectively.

Does that capture it right?

I'm wondering if it would be feasible to keep the Controller scripts fully separate from standard Zeek functionality, something like this:

- Move `base/frameworks/cluster/controller` into `policy/frameworks/cluster/controller`.
- Remove `@load base/frameworks/cluster/{agent, controller}` from `init-supervisor.zeek.`
- Instead have the top-level scripts like `test-a2c.zeek` include `@load frameworks/cluster/{agent, controller}` depending on what they need.
- Remove the `ClusterController::enable` flag, just loading scripts gets it enabled.
- And maybe: let `zeek -j` also load `init-default`, and use `zeek -jb` to run with minimal, bare supervisor-only. "maybe" because the core of the question here is what "running a supervisor" means: is it a normal Zeek with additional functionality, or is it instead something new that should really be bootstrapping freshly from a minimal configuration? My gut feeling is that the former blends in more naturally with how things are currently, but not fully sure.

Would this work? Not sure if I'm missing something in how scripts gets passed to supervised processed.

One thing to keep in mind is that eventually the controller functionality will usually be spawned from something like `systemd`, so it doesn't matter that if the Zeek command line is super-pretty. In other words, I don't think we need to have `-j` do a lot of magic by itself.

ckreibich commented on scripts/base/frameworks/supervisor/main.zeek:
```diff
@@ -42,6 +41,13 @@ function Supervisor::node(): Supervisor::NodeConfig
 
 event zeek_init() &priority=10
 	{
+	if ( Supervisor::is_supervisor() && SupervisorControl::enable_listen )
```
The difference here is that without loading the (existing) cluster framework, the supervisor process doesn't actually listen. I had to add that here explicitly.

ckreibich commented:
Thanks Robin. To confirm: your understanding of the bootstrapping logic is correct.

It's fine with me to have to load controller/agent explicitly. That's actually how it used to work. At some point I figured that if you run the supervisor, odds are you'll want the controller/agent, and I added the `enable` switches. But that might be an incorrect assumption.

I added `init-supervisor.zeek` because I figured that running the supervisor should be its own thing — not bare mode, and not (by default) be running `init-default`. If I read your comments correctly, you're on board with keeping `init-supervisor.zeek`, but reduce it to really just running the supervisor framework? I personally like this, because it gives us a defined starting point for what the supervisor runs.

ckreibich commented on src/zeek-setup.cc:
```diff
@@ -535,7 +535,11 @@ SetupResult setup(int argc, char** argv, Options* zopts)
+		// of the data cluster -- those nodes establish an
+		// environment variable ZEEK_CLUSTER_MGMT_NODE that defines
+		// their role (so far: controller or agent).
+		if ( options.supervisor_mode || getenv("ZEEK_CLUSTER_MGMT_NODE") )
```
Yeah, I'm not really happy with that part either. The idea here is that the supervisor-level agent/controller code tell the created node what role it serves. Similarly to the question of what the supervisor itself runs, I didn't want the forked node to load all scripts — I figured the agent process just need to load the agent scripts, etc.

When one tells the supervisor to create a node, one can tell it to load specific scripts _in addition_ to what it would normally load (which would be `init-default`). I considered using that, but it doesn't give me a way to say what _not_ to load.

One thought would be to convey the init script itself via the environment. I.e., if you set `ZEEK_INIT_SCRIPT` to `init-default`, Zeek will load that, if you say `init-supervisor`, it'll be that, etc. It'd keep the cluster management out of it.

ckreibich commented:
Robin and I talked this one over and we settled on the following:
- Running with `-j` will henceforth imply that you're running with a Supevisor-specific script constelltion, expressed in `init-supervisor.zeek`. Initially that will just be the supervisor framework, but could broaden in the future. This is semantically different from bare mode.
- Running the supervisor should not imply running agent & controller, so I'll move them out into policy.
- To give the supervisor API a way to state that a forked node should start from a minimal setup, we'll add a bit to `NodeConfig` that states to launch the node in bare mode. We considered going all the way to overriding the Zeek command line args, but that's a bigger change and potentially troublesome if the flags aren't chosen carefully.
- We can remove the supervisor-specific configurability of address and port it listens on, since just using the Broker-level settings is fine — there's no need for any process to listen on more than one port. The on/off bit for listening stays, though.

ckreibich commented:
Okay Robin, I think this covers it. The last few pushes were iterations on the testcase, where I noticed some instability. I've updated https://github.com/ckreibich/cc-exploration so it works with the new approach. I like it a lot.

We haven't resolve the main.zeek naming question — I could still do that, but I think that could also happen later. It's a bit fiddly to adjust naming of files under `policy/` because of those btests that check the list of files. There's also no NEWS entry yet.

rsmmr commented on scripts/base/frameworks/supervisor/main.zeek:
```diff
@@ -42,6 +41,13 @@ function Supervisor::node(): Supervisor::NodeConfig
 	{
+	if ( Supervisor::is_supervisor() && SupervisorControl::enable_listen )
+		{
+		Broker::listen(Broker::default_listen_address,
```
Looks good, one more nit: maybe this could move into a Broker-side function now, `Broker::listen_default()` or so.

ckreibich replied:
Nice one, yeah.

rsmmr reviewed (approved):
This is great now, ready to merge I'd say. We will iterate on the main scripts more later anyways, but this looks like a nice way to integrate the new functionality. 

(And fIne to stay with `main.zeek`.)

ckreibich commented:
I'm trying to understand the sanitizer failure and will merge once that's sorted.

ckreibich commented:
The sanitizer failure indirectly highlighted an oversight in the introduction of the bare-mode flag: since it defaulted to `F`, it caused the bare-mode status of a Supervisor running with it to get lost, loading the new node with the full set of scripts, which translates into much longer test run times that tripped up the timeout. It also meant we changed behavior where we arguably shouldn't.

I've changed the feature so the flag is `&optional`, so when it's not present new nodes inherit the bare-mode status of the Supervisor.

The second force-push was a rebase on top of master.
